src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/DefaultTagHelperContent.cs

```diff
@@ -82,73 +94,37 @@ public override bool IsWhiteSpace
         }
 
         /// <inheritdoc />
-        public override bool IsEmpty
+        public override TagHelperContent Append(string unencoded) => AppendCore(unencoded);
+
+        /// <inheritdoc />
+        public override TagHelperContent AppendHtml(string encoded)
         {
-            get
+            if (encoded == null)
             {
-                if (!IsModified)
-                {
-                    return true;
-                }
+                return AppendCore(null);
+            }
 
-                using (var writer = new EmptyOrWhiteSpaceWriter())
-                {
-                    foreach (var entry in _buffer)
-                    {
-                        if (entry == null)
-                        {
-                            continue;
-                        }
+            return AppendCore(new HtmlEncodedString(encoded));
+        }
 
-                        var stringValue = entry as string;
-                        if (stringValue != null)
-                        {
-                            if (!string.IsNullOrEmpty(stringValue))
-                            {
-                                return false;
-                            }
-                        }
-                        else
-                        {
-                            ((IHtmlContent)entry).WriteTo(writer, HtmlEncoder.Default);
-                            if (!writer.IsEmpty)
-                            {
-                                return false;
-                            }
-                        }
-                    }
-                }
+        /// <inheritdoc />
+        public override TagHelperContent AppendHtml(IHtmlContent htmlContent) => AppendCore(htmlContent);
```

Comment:
Since you said offline that consistency was one of your explicit goals :smiling_imp:, move this above `AppendHtml(string)`.